File: 프로그래머스/3/42579. 베스트앨범/베스트앨범.py
def solution(genres, plays):
    answer = []
    index_genres_plays_dic = {}
    play_amount_dic = {}

    for i, (g, p) in enumerate(zip(genres, plays)):
        if g not in index_genres_plays_dic:
            index_genres_plays_dic[g] = [(i, p)]
        else:
            index_genres_plays_dic[g].append((i,p))
    
        if g not in play_amount_dic:
            play_amount_dic[g] = p

        else:
            play_amount_dic[g] += p

    for (key, value) in sorted(play_amount_dic.items(), key=lambda x:x[1], reverse=True):
        for (index, play) in sorted(index_genres_plays_dic[key], key=lambda x:x[1], reverse=True)[:2]:
            answer.append(index)

    return answer

File: 프로그래머스/3/42579. 베스트앨범/베스트앨범.cpp
#include <string>
#include <vector>
#include <unordered_map>
#include <algorithm>

using namespace std;

vector<int> solution(vector<string> genres, vector<int> plays) {
    
    unordered_map<string, int> genre_total;
    unordered_map<string, vector<pair<int, int>>> genre_songs;
    
    for (int i=0; i<genres.size(); i++) {
        genre_total[genres[i]] += plays[i];
        genre_songs[genres[i]].push_back({plays[i], i});
    }
    
    vector<pair<int, string>> genre_order;
    for (auto it: genre_total) {
        genre_order.push_back({it.second, it.first});
    }
    sort(genre_order.rbegin(), genre_order.rend());
    
    vector<int> answer;
    
    for (auto g: genre_order) {
        string genre = g.second;
        auto &songs = genre_songs[genre];
        
        sort(songs.begin(), songs.end(), [](pair<int, int>a, pair<int, int>b){
            if (a.first != b.first) return a.first > b.first;
            else return a.second < b.second;
        });
        for (int i=0; i<songs.size() && i < 2; i++) {
            answer.push_back(songs[i].second);
        }
    }
    
    return answer;
}